z3/src/ast.rs

```diff
@@ -1237,38 +1238,97 @@ impl<'ctx> Datatype<'ctx> {
 ///
 /// let x = ast::Int::new_const(&ctx, "x");
 /// let f_x: ast::Int = f.apply(&[&x.clone().into()]).try_into().unwrap();
-/// let forall: ast::Dynamic = ast::forall_const(&ctx, &[&x.clone().into()], &(x._eq(&f_x)).into());
-/// solver.assert(&forall.try_into().unwrap());
+/// let f_x_pattern: Pattern = Pattern::new(&ctx, &[ &f_x.clone().into() ]);
+/// let forall: ast::Bool = ast::forall_const(
+///     &ctx,
+///     &[&x.clone().into()],
+///     &[&f_x_pattern],
+///     &x._eq(&f_x).into()
+/// ).try_into().unwrap();
+/// solver.assert(&forall);
 ///
 /// assert!(solver.check());
 /// let model = solver.get_model();
 ///
 /// let f_f_3: ast::Int = f.apply(&[&f.apply(&[&ast::Int::from_u64(&ctx, 3).into()])]).try_into().unwrap();
 /// assert_eq!(3, model.eval(&f_f_3).unwrap().as_u64().unwrap());
 /// ```
-
 pub fn forall_const<'ctx>(
     ctx: &'ctx Context,
     bounds: &[&Dynamic<'ctx>],
+    patterns: &[&Pattern<'ctx>],
     body: &Dynamic<'ctx>,
 ) -> Dynamic<'ctx> {
-    assert!(bounds.iter().all(|a| a.get_ctx().z3_ctx == ctx.z3_ctx));
+    assert!(bounds.iter().all(|a| a.get_ctx() == ctx));
+    assert!(patterns.iter().all(|p| p.ctx == ctx));
     assert_eq!(ctx, body.get_ctx());
 
-    if bounds.is_empty() {
-        return body.clone();
-    }
-
```

Comment:
I removed this because I would prefer to avoid adding optimizations to this crate (and because Z3 should be able to easily optimize this case). Let me know if you have a different vision.

Reply:
It might be good to make sure that Z3 actually does optimize this ... Either way, if this is removed, I'd rather see it happen in a separate PR.